### PR TITLE
[charlie] proposed hotfix: LMDB reader lock exhaustion hack

### PR DESCRIPTION
**Pull request by marc-casperlabs**
This is a potential quick fix for the LMDB reader issue. It's not something we want to keep around long term, but it might solve the problem (@momipsl can you give this a whirl maybe, if you can reproduce the actual issue reliably?)

It works by retrying failed reads with a random backoff. This will cause a pile-up, but we don't have an unbounded number of reads to do, so this is essentially poor man's queuing through spawning extra threads.

### CODE REVIEW

**Comment by goral09** [comment]
I think another fix, and simpler b/c doesn't change our storage access story, is to cache the `Chainspec` instead of fetching it for every incoming deploy (and for every block being executed).

File: node/src/components/storage/lmdb_store.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-use std::{fmt::Debug, marker::PhantomData, path::Path};
+use std::{fmt::Debug, marker::PhantomData, path::Path, thread::sleep, time::Duration};
```
Comment (Fraser999) [inline node/src/components/storage/lmdb_store.rs:1]:
```suggestion
use std::{fmt::Debug, marker::PhantomData, path::Path, thread, time::Duration};
```

File: node/src/components/storage/lmdb_store.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ use smallvec::smallvec;
 use tracing::info;
 
 use super::{Error, Multiple, Result, Store, Value};
+use rand::{thread_rng, Rng};
```
Comment (Fraser999) [inline node/src/components/storage/lmdb_store.rs:10]:
```suggestion
use rand::{self, Rng};
```

**Review by Fraser999 — CHANGES_REQUESTED** [review]
I'm not sure this piling up of reads will actually help or just mutate the problem into one of resource consumption.  I think there are better quick fixes than this.

At the very least, this change should be applied to the `LmdbChainspecStore`, since that's been the one causing the panics.

**Comment by marc-casperlabs** [comment]
> I'm not sure this piling up of reads will actually help or just mutate the problem into one of resource consumption. I think there are better quick fixes than this.

No doubt. The question is can we produce one for master that backports cleanly onto charlie?

> At the very least, this change should be applied to the LmdbChainspecStore, since that's been the one causing the panics.

True. Should be we apply it to both?

**Comment by EdHastingsCasperLabs** [comment]
While I appreciate the initiative taken here to offer a potential hotfix for the issue observed in charlie and understand that it is meant very much as a hotfix and not as a real solution. 

However, I favor the more directed approach taken in https://github.com/CasperLabs/casper-node/pull/281 as in addition to hotfix the symptom observed in the test branch, I think it is also a beneficial enhancement to master.

While it is possible that one of the other lmdb stores will fail in the same way once the chainspec usage is no longer panicking, I think that is beneficial in the current timeframe as it would give us useful detection of hotspots in our data consumption patterns and allow us to make further informed and targeted correction.

**Comment by marc-casperlabs** [comment]
I'm fine with closing this, but ensuring we don't exhaust the reader pool is orthogonal to any other fix still (@Fraser999 should take a look at the issue).

**Pull request closed without merging**